File: blue-core/src/blue/noteProcessor/RetrogradeProcessor.java
package blue.noteProcessor;

import blue.plugin.NoteProcessorPlugin;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;

/**
 * <p>
 * Title: blue
 * </p>
 * <p>
 * Description: an object composition environment for csound
 * </p>
 * <p>
 * Copyright: Copyright (c) 2001
 * </p>
 * <p>
 * Company: steven yi music
 * </p>
 * 
 * @author steven yi
 * @version 1.0
 */

@NoteProcessorPlugin(displayName="RetrogradeProcessor", position = 80)
public class RetrogradeProcessor implements NoteProcessor, java.io.Serializable {

    public RetrogradeProcessor() {
    }

    @Override
    public String toString() {
        return "[retrograde]";
    }

    public final void processNotes(NoteList in) {
        in.sort();
        Note temp;
        int size = in.size();
        temp = in.get(in.size() - 1);
        float totalTime = temp.getStartTime() + temp.getSubjectiveDuration();

        for (int i = 0; i < size; i++) {
            temp = in.get(i);
            // System.out.println("obj: " + temp.getObjectiveDuration() + "
            // subj: " + temp.getSubjectiveDuration());
            temp.setStartTime(totalTime
                    - (temp.getStartTime() + temp.getSubjectiveDuration()));
        }
    }

    public static void main(String[] args) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.add(Note.createNote("i1 " + i + " " + i + " 6." + i
                        + " 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        RetrogradeProcessor retro = new RetrogradeProcessor();
        retro.processNotes(n);

        System.out.println("after: \n\n" + n + "\n\n");

    }

    public static NoteProcessor loadFromXML(Element data) {
        return new RetrogradeProcessor();
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        return retVal;
    }
}

File: packages/blue-data/src/note-processors/retrograde-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';

const JAVA_TYPE = 'blue.noteProcessor.RetrogradeProcessor';

export class RetrogradeProcessor extends NoteProcessor {
  constructor();
  constructor(_src: RetrogradeProcessor);
  constructor(_src?: RetrogradeProcessor) {
    super();
  }

  override process(notes: NoteList): NoteList {
    notes.sort();
    const size = notes.length;
    if (size === 0) return notes;

    const lastNote = notes.getNote(size - 1);
    const totalTime = lastNote.getStartTime() + lastNote.getSubjectiveDuration();

    for (let i = 0; i < size; i++) {
      const temp = notes.getNote(i);
      temp.setStartTime(totalTime - (temp.getStartTime() + temp.getSubjectiveDuration()));
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'RetrogradeProcessor';
  }

  override deepCopy(): RetrogradeProcessor {
    return new RetrogradeProcessor();
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    return elem;
  }

  static loadFromXML(_data: Element): RetrogradeProcessor {
    return new RetrogradeProcessor();
  }
}
